tensorflow/java/src/main/java/org/tensorflow/Tensor.java

```diff
@@ -537,56 +548,72 @@ private static void throwExceptionIfNotByteOfByteArrays(Object array) {
+    class_datatypes.put(Long.class, DataType.INT64);
+    class_datatypes.put(float.class, DataType.FLOAT);
+    class_datatypes.put(Float.class, DataType.FLOAT);
+    class_datatypes.put(double.class, DataType.DOUBLE);
+    class_datatypes.put(Double.class, DataType.DOUBLE);
+    class_datatypes.put(byte.class, DataType.STRING);
+    class_datatypes.put(Byte.class, DataType.STRING);
+    class_datatypes.put(boolean.class, DataType.BOOL);
+    class_datatypes.put(Boolean.class, DataType.BOOL);
+  }
+
   private static DataType dataTypeOf(Object o) {
-    if (o.getClass().isArray()) {
-      if (Array.getLength(o) == 0) {
-        throw new IllegalArgumentException("cannot create Tensors with a 0 dimension");
-      }
-      // byte[] is a DataType.STRING scalar.
-      Object e = Array.get(o, 0);
-      if (e == null) {
-        throwExceptionIfNotByteOfByteArrays(o);
-        return DataType.STRING;
-      }
-      if (Byte.class.isInstance(e) || byte.class.isInstance(e)) {
-        return DataType.STRING;
-      }
-      return dataTypeOf(e);
+    Class<?> c = o.getClass();
+    while (c.isArray()) {
+      c = c.getComponentType();
     }
-    if (Float.class.isInstance(o) || float.class.isInstance(o)) {
-      return DataType.FLOAT;
-    } else if (Double.class.isInstance(o) || double.class.isInstance(o)) {
-      return DataType.DOUBLE;
-    } else if (Integer.class.isInstance(o) || int.class.isInstance(o)) {
-      return DataType.INT32;
-    } else if (Long.class.isInstance(o) || long.class.isInstance(o)) {
-      return DataType.INT64;
-    } else if (Boolean.class.isInstance(o) || boolean.class.isInstance(o)) {
-      return DataType.BOOL;
-    } else {
-      throw new IllegalArgumentException("cannot create Tensors of " + o.getClass().getName());
+    DataType ret = class_datatypes.get(c);
+    if (ret != null) { 
+      return ret;
     }
+    throw new IllegalArgumentException("cannot create Tensors of type " + c.getName());
   }
 
-  private static int numDimensions(Object o) {
-    if (o.getClass().isArray()) {
-      Object e = Array.get(o, 0);
-      if (e == null) {
-        throwExceptionIfNotByteOfByteArrays(o);
-        return 1;
-      } else if (Byte.class.isInstance(e) || byte.class.isInstance(e)) {
-        return 0;
-      }
-      return 1 + numDimensions(e);
+  /** Returns the number of dimensions of a tensor of type dtype
+   *  when represented by the object o.
+   */
+  private static int numDimensions(Object o, DataType dtype) {
+    int ret = numArrayDimensions(o);
+    if (dtype == DataType.STRING && ret > 0) {
+      return ret - 1;
+    }
+    return ret;
+  }
+  
+  /** Returns the number of dimensions of the array object o.
+   *  Returns 0 if o is not an array.
+   */
+  private static int numArrayDimensions(Object o) {
+    Class<?> c = o.getClass();
+    if (!c.isArray()) return 0;
+    String s = c.getName();
+    int i = 0;
+    while (s.charAt(i) == '[') {
```

Comment:
This could also be done by walking up the `getComponentType()` path but I'm conjecturing it's a little faster to just walk through the string.

It should also be correct to initialize i to 1 instead of 0 since the first character is guaranteed to be `[`, but starting from zero seemed more obviously correct .

Reply:
I don't have a strong opinion either way (though is there any reason to believe that one would be significantly faster than the other?)